### ISSUE TITLE
Only a single error is emitted for multiple equal unmet bounds arising from distinct type parameters on the same item

### ISSUE THREAD

**Issue by fmease** [rust-lang/rust#115288]
Uplifted from https://github.com/rust-lang/rust/issues/105306#issuecomment-1518136153 & https://github.com/rust-lang/rust/issues/105306#issuecomment-1693645473.
CC @estebank

For the following erroneous code, rustc only emits a single error while ideally it should emit two errors. Presumably, this happens because the two predicate obligations `String: Copy` have the same cause span (namely of the entire path) before any adjustment takes place and therefore get deduplicated early on. Details may vary.

Original reproducer (involving projections):

```rs
trait Trait {
    type P<T: Copy, U: Copy>;
}
impl Trait for () {
    type P<T: Copy, U: Copy> = ();
}
fn main() {
    let _: <() as Trait>::P<String, String>;
}
```

Current output (1 error):

```
error[E0277]: the trait bound `String: Copy` is not satisfied
 --> src/main.rs:8:29
  |
8 |     let _: <() as Trait>::P<String, String>;
  |                             ^^^^^^ the trait `Copy` is not implemented for `String`
  |
note: required by a bound in `Trait::P`
 --> src/main.rs:2:15
  |
2 |     type P<T: Copy, U: Copy>;
  |               ^^^^ required by this bound in `Trait::P`
```

Expected output (2 errors):

```
error[E0277]: the trait bound `String: Copy` is not satisfied
 --> src/main.rs:8:12
  |
8 |     let _: <() as Trait>::P<String, String>;
  |                             ^^^^^^ the trait `Copy` is not implemented for `String`
  |
note: required by a bound in `Trait::P`
 --> src/main.rs:2:15
  |
2 |     type P<T: Copy, U: Copy>;
  |               ^^^^ required by this bound in `Trait::P`

error[E0277]: the trait bound `String: Copy` is not satisfied
 --> src/main.rs:8:12
  |
8 |     let _: <() as Trait>::P<String, String>;
  |                                     ^^^^^^ the trait `Copy` is not implemented for `String`
  |
note: required by a bound in `Trait::P`
 --> src/main.rs:2:24
  |
2 |     type P<T: Copy, U: Copy>;
  |                        ^^^^ required by this bound in `Trait::P`

```

Another very similar reproducer (involving ADTs) but it probably takes a slightly different code path in the compiler:

```rs
struct S<T: Copy, U: Copy>(T, U);

fn main() {
    let _: S<String, String>;
}
```

### Meta

<details><summary><code>rustc -Vv</code></summary>

```
rustc 1.74.0-nightly (668bf8c59 2023-08-27)
binary: rustc
commit-hash: 668bf8c5932933255ba278f50fdbe308f7abe60f
commit-date: 2023-08-27
host: x86_64-unknown-linux-gnu
release: 1.74.0-nightly
LLVM version: 17.0.0
```

</details>

@rustbot label C-bug T-compiler A-diagnostics D-incorrect
